File: Sources/SplitLaneCore/Proxy/SOCKS5/ByteReader.swift
import Foundation

/// A bounds-checked forward cursor over a byte buffer.
///
/// Every byte the SOCKS5 layer reads from the upstream goes through this type. Nothing in that
/// layer subscripts a received buffer directly, and that is a security requirement rather than a
/// style preference: Swift *traps* on out-of-range slicing, and a trap inside the system extension
/// kills the provider, which fails every selected application open until it restarts. A malformed
/// reply from a hostile or MITM'd proxy must produce an error, never a crash. See F-7 in
/// docs/THREAT_MODEL.md.
///
/// Reads are transactional at the call site: a read that cannot be satisfied throws
/// ``SOCKS5Error/incompleteResponse`` and leaves the cursor untouched, so a caller parsing a
/// partially arrived message can retry against the same reader once more bytes land.
struct ByteReader {

    private let bytes: [UInt8]
    private(set) var offset: Int

    init(_ data: Data) {
        self.bytes = [UInt8](data)
        self.offset = 0
    }

    init(_ bytes: [UInt8]) {
        self.bytes = bytes
        self.offset = 0
    }

    /// Bytes not yet consumed.
    var remaining: Int { bytes.count - offset }

    var isAtEnd: Bool { remaining == 0 }

    /// Reads one byte.
    mutating func readUInt8() throws -> UInt8 {
        guard remaining >= 1 else { throw SOCKS5Error.incompleteResponse }
        defer { offset += 1 }
        return bytes[offset]
    }

    /// Reads a big-endian (network byte order) 16-bit value.
    mutating func readUInt16() throws -> UInt16 {
        guard remaining >= 2 else { throw SOCKS5Error.incompleteResponse }
        defer { offset += 2 }
        return UInt16(bytes[offset]) << 8 | UInt16(bytes[offset + 1])
    }

    /// Reads exactly `count` bytes.
    ///
    /// `count` is frequently attacker-influenced — the `DOMAIN` length byte of a reply is a value
    /// the upstream chooses — so the length check is the whole point of this method.
    mutating func readBytes(_ count: Int) throws -> [UInt8] {
        guard count >= 0 else { throw SOCKS5Error.malformedResponse("negative length") }
        guard remaining >= count else { throw SOCKS5Error.incompleteResponse }
        defer { offset += count }
        return Array(bytes[offset..<(offset + count)])
    }

    /// Reads a length-prefixed byte string (one length byte followed by that many bytes).
    mutating func readLengthPrefixedBytes() throws -> [UInt8] {
        let length = try readUInt8()
        return try readBytes(Int(length))
    }

    /// Peeks without consuming.
    func peekUInt8(at index: Int = 0) -> UInt8? {
        let position = offset + index
        guard position >= 0, position < bytes.count else { return nil }
        return bytes[position]
    }
}

File: windows/src/SplitLane.Core/Proxy/Socks5/ByteReader.cs
namespace SplitLane.Core.Proxy.Socks5;

/// <summary>
/// A bounds-checked cursor over a buffer of received bytes.
/// </summary>
/// <remarks>
/// Every read either succeeds or throws <see cref="Socks5ErrorCode.IncompleteResponse"/>, and the
/// offset only advances on success. That contract is what lets the negotiator re-parse a message
/// from the top each time more bytes arrive without tracking partial state: a truncated read is a
/// signal to wait, not a failure, and it cannot leave the reader half-consumed.
///
/// <para>
/// It is a <c>ref struct</c> so it cannot outlive or escape the buffer it points at, and so parsing
/// a handshake allocates nothing at all.
/// </para>
/// </remarks>
internal ref struct ByteReader(ReadOnlySpan<byte> buffer)
{
    private readonly ReadOnlySpan<byte> _buffer = buffer;

    /// <summary>How many bytes have been consumed.</summary>
    public int Offset { get; private set; }

    /// <summary>Bytes not yet consumed.</summary>
    public readonly int Remaining => _buffer.Length - Offset;

    /// <summary>Reads one byte.</summary>
    public byte ReadUInt8()
    {
        if (Remaining < 1)
        {
            throw Socks5Exception.Incomplete();
        }

        return _buffer[Offset++];
    }

    /// <summary>Reads a big-endian 16-bit value, which is the only width SOCKS5 uses for ports.</summary>
    public ushort ReadUInt16()
    {
        if (Remaining < 2)
        {
            throw Socks5Exception.Incomplete();
        }

        var value = (ushort)((_buffer[Offset] << 8) | _buffer[Offset + 1]);
        Offset += 2;
        return value;
    }

    /// <summary>Reads a fixed number of bytes.</summary>
    public byte[] ReadBytes(int count)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(count);

        if (Remaining < count)
        {
            throw Socks5Exception.Incomplete();
        }

        var slice = _buffer.Slice(Offset, count).ToArray();
        Offset += count;
        return slice;
    }

    /// <summary>
    /// Reads a single-byte length followed by that many bytes.
    /// </summary>
    /// <remarks>
    /// The length is chosen by the server. Because <see cref="ReadBytes"/> bounds-checks it against
    /// what actually arrived, a hostile length can only produce an "incomplete" signal — never an
    /// over-read, and never an allocation the server controls the size of beyond 255 bytes.
    /// </remarks>
    public byte[] ReadLengthPrefixedBytes()
    {
        var length = ReadUInt8();
        return ReadBytes(length);
    }
}
